### PR TITLE
[webview_flutter] Implementation of the webview_flutter_platform_interface package

### PR DESCRIPTION
**Pull request by mvanbeusekom**
This PR adds the webview_flutter_platform_interface package which can be used by plugin developers to add platform support to the webview_flutter plugin.

flutter/flutter#86286

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

File: packages/webview_flutter/webview_flutter_platform_interface/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## 1.0.0
+
+* Initial open-source release.
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_platform_interface/CHANGELOG.md:3]: This description doesn't really make sense; how about "Extracted platform interface from `webview_flutter`"

File: packages/webview_flutter/webview_flutter_platform_interface/test/src/method_channel/webview_method_channel_test.dart [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+import 'package:mockito/mockito.dart';
+
+import 'package:webview_flutter_platform_interface/src/method_channel/webview_method_channel.dart';
+import 'package:webview_flutter_platform_interface/src/utils/javascript_channel_registry.dart';
+import 'package:webview_flutter_platform_interface/webview_flutter_platform_interface.dart';
+
+void main() {
+  TestWidgetsFlutterBinding.ensureInitialized();
+
+  group('$MethodChannelWebViewPlatform', () {
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_platform_interface/test/src/method_channel/webview_method_channel_test.dart:16]: I don't know why the plugins repo does so much of this; tests are already grouped by the file they are in. I'd rather just not have a group wrapping all tests. (Logic groups of sets of tests within the file are fine, it's just these top-level ones that seem like noise.)

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
A few things from a first pass:
- This is a big enough PR that I'd rather it not do two things at once; let's make moving to a subfolder its own separate precursor PR. (That's actually an example of something that would fail the new check I'm adding.)
- There should be an AUTHORS file for the new package (filed https://github.com/flutter/flutter/issues/89680), cloned from the main package as a starting point.
- Because it's a copy rather than a move it's hard to know what (if anything) has changed. Could you separate out a first commit that is *just* a straight copy of all the relevant files to the new location, then do any changes in separate commits? I see that there's some of that, but, e.g., the new CHANGELOG is in that commit so I know it's not a pure copy. I can't easily see what other actual changes are there to review.

File: packages/webview_flutter/webview_flutter_platform_interface/CHANGELOG.md (hunk 1 above)
Comment (mvanbeusekom) [reply packages/webview_flutter/webview_flutter_platform_interface/CHANGELOG.md]: Much better, thank you. I will pay more attention to this in the future.

File: packages/webview_flutter/webview_flutter_platform_interface/test/src/method_channel/webview_method_channel_test.dart (hunk 2 above)
Comment (mvanbeusekom) [reply packages/webview_flutter/webview_flutter_platform_interface/test/src/method_channel/webview_method_channel_test.dart]: I have updated the description so the grouping makes more sense. At the moment there are 2 groups:

- One that tests the API on the `plugin.flutter.io/webview_<channel_id>` method channel;
- A second group that tests the API on the `plugins.flutter.io/cookie_manager` method channel.

**Comment by mvanbeusekom** [comment]
> A few things from a first pass:
> 
> * This is a big enough PR that I'd rather it not do two things at once; let's make moving to a subfolder its own separate precursor PR. (That's actually an example of something that would fail the new check I'm adding.)
> * There should be an AUTHORS file for the new package (filed [[flutter_plugin_tools] Validate that AUTHORS exists flutter#89680](https://github.com/flutter/flutter/issues/89680)), cloned from the main package as a starting point.
> * Because it's a copy rather than a move it's hard to know what (if anything) has changed. Could you separate out a first commit that is _just_ a straight copy of all the relevant files to the new location, then do any changes in separate commits? I see that there's some of that, but, e.g., the new CHANGELOG is in that commit so I know it's not a pure copy. I can't easily see what other actual changes are there to review.

@stuartmorgan I have redone the pull request and divided it into 9 logical commits and added additional explanation with each commit to explain the changes made. 

I have made al changes based on copied files I done in commit bc79ecdbf0cf82976304264e6e2973a645d94825 which should make it easier for you to follow the changes.

One more thing, to not make to many changes I left the code as-is and did not update it according to the latest `analysis_options.yaml` (meaning at the moment it still relies on the `analysis_options_legacy.yaml` configuration). If you would like me to update this also as part of this PR please let me know, otherwise I will file an issue and create a new PR later on.

I would appreciate it if you could have another look.

**Comment by stuartmorgan** [comment]
> One more thing, to not make to many changes I left the code as-is and did not update it according to the latest `analysis_options.yaml` (meaning at the moment it still relies on the `analysis_options_legacy.yaml` configuration).

Yes, that should definitely be separate. Thanks!

**Review by stuartmorgan — APPROVED** [review]
Thanks for tweaking the commits; much easier to review. LGTM modulo some nits, and getting the `format` bots green.

File: packages/webview_flutter/webview_flutter_platform_interface/test/src/method_channel/webview_method_channel_test.dart [hunk 3]
```diff
@@ -0,0 +1,451 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (mvanbeusekom) [reply packages/webview_flutter/webview_flutter_platform_interface/test/src/method_channel/webview_method_channel_test.dart:1]: That is my mistake. I try to keep the directory structure the test is in as close as possible to the directory structure to the file containing the code it is testing. 

So in this case when I need to write tests for code that is in the `lib/src/types/javascript_channel.dart` I create a test file at `test/src/types/javascript_channel_test.dart`. It helps me navigate to the correct test file if I am looking to extend certain tests for an existing file.

File: packages/webview_flutter/webview_flutter_platform_interface/test/src/method_channel/webview_method_channel_test.dart [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+import 'package:webview_flutter_platform_interface/src/method_channel/webview_method_channel.dart';
+import 'package:webview_flutter_platform_interface/webview_flutter_platform_interface.dart';
+
+void main() {
+  TestWidgetsFlutterBinding.ensureInitialized();
+
+  group('Tests on `plugin.flutter.io/webview_<channel_id>` channel', () {
+    const int channelId = 1;
+    const MethodChannel channel =
+    MethodChannel('plugins.flutter.io/webview_$channelId');
```
[the remark this answers is not in the document]
Comment (mvanbeusekom) [reply packages/webview_flutter/webview_flutter_platform_interface/test/src/method_channel/webview_method_channel_test.dart]: Run the auto-formatter so formatting steps should go green soon.

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_interface/webview_platform_controller.dart [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  /// Return the vertical scroll position of this view.
+  ///
+  /// Scroll position is measured from top.
+  Future<int> getScrollY() {
+    throw UnimplementedError(
+        "WebView getScrollY is not implemented on the current platform");
+  }
+}
```
[the remark this answers is not in the document]
Comment (mvanbeusekom) [reply packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_interface/webview_platform_controller.dart]: Did a full sweep, thanks for pointing it out.

File: packages/webview_flutter/webview_flutter_platform_interface/README.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+# Note on breaking changes
+
+Strongly prefer non-breaking changes (such as adding a method to the interface)
+over breaking changes for this package.
+
+See https://flutter.dev/go/platform-interface-breaking-changes for a discussion
+on why a less-clean interface is preferable to a breaking change.
+
+[1]: ../webview_flutter
```
[the remark this answers is not in the document]
Comment (mvanbeusekom) [reply packages/webview_flutter/webview_flutter_platform_interface/README.md]: Fixed.

**Comment by stuartmorgan** [comment]
Oh, I didn't consider this effect of my change to the Windows bots... the two "Windows Plugins" jobs are never going to run because the builders no longer exist for them (I renamed them). You'll need to sync in the last master.

**Comment by mvanbeusekom** [comment]
> Oh, I didn't consider this effect of my change to the Windows bots... the two "Windows Plugins" jobs are never going to run because the builders no longer exist for them (I renamed them). You'll need to sync in the last master.

No worries, will do in a few minutes.

**Pull request merged**